frontend/catalyst/api_extensions/control_flow.py

```diff
@@ -316,17 +321,62 @@ def loop_rx(i, x):
 
     >>> circuit(7, 1.6)
     [array(0.97926626), array(0.55395718)]
+
+    Note that using dynamically-shaped arrays within for loops, while loops, and
+    conditional statements, are also supported:
+
+    >>> @qjit
+    ... def f(shape):
+    ...     a = jnp.ones([shape], dtype=float)
+    ...     @for_loop(0, 10, 2)
+    ...     def loop(i, a):
+    ...         return a + i
+    ...     return loop(a)
+    >>> f(5)
+    array([21., 21., 21., 21., 21.])
+
+    By default, ``allow_array_resizing`` is ``False``, allowing dynamical-shaped
+    arrays from outside the for loop to be correctly captured:
+
+    >>> @qjit(abstracted_axes={1: 'n'})
+    ... def g(x, y):
+    ...     @catalyst.for_loop(0, 10, 1)
+    ...     def loop(_, a):
+    ...         # Attempt to capture `x` from the outer scope.
+    ...         return a * x
+    ...     return jnp.sum(loop(y))
+    >>> a = jnp.ones([1,3], dtype=float)
+    >>> b = jnp.ones([1,3], dtype=float)
+    >>> g(a, b)
+    array(3.)
+
+    However, if you wish to have the for loop return differently sized/shaped arrays
+    at each iteration, set ``allow_array_resizing`` to ``True``:
+
+    >>> @qjit()
+    ... def f(N):
+    ...     a = jnp.ones([N], dtype=float)
+    ...     @for_loop(0, 10, 1, allow_array_resizing=True)
+    ...     def loop(i, _):
+    ...         return jnp.ones([i], dtype=float) # return array of new dimensions
+    ...     return loop(a)
+    >>> f(5)
+    array([1., 1., 1., 1., 1., 1., 1., 1., 1.])
+
+    Note that when ``allow_array_resizing=True``, dynamically-shaped arrays
+    can no longer be captured from outer-scopes by the for loop. For more details
+    on dynamically-shaped arrays, please see :ref:`dynamic-arrays`.
     """
 
     def _decorator(body_fn):
         return ForLoopCallable(
-            lower_bound, upper_bound, step, body_fn, experimental_preserve_dimensions
+            lower_bound, upper_bound, step, body_fn, not allow_array_resizing
```

Comment:
Not ideal, but this avoided needing to significantly change the code in this file